README.md
# Python 小程式範例

## scrap1.py -> 爬蟲程式並將結果輸出至index.csv

## tts.py tts3.py -> 訊飛 文字轉音頻 WEB API
## tts2.py -> 訊飛 文字轉音頻 SDK
## word.py word5.py -> 訊飛 音頻轉文字 WEB API


news.txt
{"code":"0","data":"请说出你的身高和体重。","desc":"success","sid":"zat00381915@ch084b0e3a4248515100"}

tts.py
#訊飛文字轉音檔WEB_API版
#! python3
import base64
import json
import time
import hashlib
import urllib.request
import urllib.parse

# API请求地址、API KEY、APP ID等参数，提前填好备用
api_url = "http://api.xfyun.cn/v1/service/v1/tts"
#API_KEY = "f24968286f853d6138ff1fda9f1bce10"
API_KEY = "f24968286f853d6138ff1fda9f1bce10"
APP_ID = "5ade9bcc"
OUTPUT_FILE = "d:/python/output.mp3"    # 输出音频的保存路径，请根据自己的情况替换
TEXT = "苟利国家生死以，岂因祸福避趋之"

# 构造输出音频配置参数
Param = {
    "auf": "audio/L16;rate=16000",    #音频采样率
    "aue": "lame",    #音频编码，raw(生成wav)或lame(生成mp3)
    "voice_name": "xiaoyan",
    "speed": "50",    #语速[0,100]
    "volume": "77",    #音量[0,100]
    "pitch": "50",    #音高[0,100]
    "engine_type": "aisound"    #引擎类型。aisound（普通效果），intp65（中文），intp65_en（英文）
}
# 配置参数编码为base64字符串，过程：字典→明文字符串→utf8编码→base64(bytes)→base64字符串
Param_str = json.dumps(Param)    #得到明文字符串
Param_utf8 = Param_str.encode('utf8')    #得到utf8编码(bytes类型)
Param_b64 = base64.b64encode(Param_utf8)    #得到base64编码(bytes类型)
Param_b64str = Param_b64.decode('utf8')    #得到base64字符串

# 构造HTTP请求的头部
time_now = str(int(time.time()))
checksum = (API_KEY + time_now + Param_b64str).encode('utf8')
checksum_md5 = hashlib.md5(checksum).hexdigest()
header = {
    "X-Appid": APP_ID,
    "X-CurTime": time_now,
    "X-Param": Param_b64str,
    "X-CheckSum": checksum_md5
}

# 构造HTTP请求Body
body = {
    "text": TEXT
}
body_urlencode = urllib.parse.urlencode(body)
body_utf8 = body_urlencode.encode('utf8')

# 发送HTTP POST请求
req = urllib.request.Request(api_url, data=body_utf8, headers=header)
response = urllib.request.urlopen(req)

# 读取结果
response_head = response.headers['Content-Type']
if(response_head == "audio/mpeg"):
    out_file = open(OUTPUT_FILE, 'wb')
    data = response.read() # a 'bytes' object
    out_file.write(data)
    out_file.close()
    print('输出文件: ' + OUTPUT_FILE)
else:
    print(response.read().decode('utf8'))

tts2.py
#訊飛文字轉音檔SDK版
#! /usr/bin/env python  
# -*- coding: utf-8 -*-  
# python version >3.5  
#! python3
__author__ = 'onestab'  
  
import sys, os, time  
from datetime import datetime  
from ctypes import *  
  
dll = windll.LoadLibrary("msc.dll")  
login_params = b"appid = 12345678, work_dir = ."  
session_begin_params = b"engine_type = local, voice_name = xiaoyan, text_encoding = UTF8, tts_res_path = fo|res\\tts\\xiaoyan.jet;fo|res\\tts\\common.jet, sample_rate = 16000, speed = 80, volume = 70, pitch = 50, rdn = 0"  
  
wav_header = b'RIFF6n\x06\x00WAVEfmt \x10\x00\x00\x00\x01\x00\x01\x00\x80>\x00\x00\x00}\x00\x00\x02\x00\x10\x00data\x00\x00\x00\x00'  
MSP_SUCCESS = 0  
MSP_TTS_FLAG_STILL_HAVE_DATA = 1  
MSP_TTS_FLAG_DATA_END = 2  
MSP_TTS_FLAG_CMD_CANCELED = 4  
  
filename = "tts_sample.wav"  
  
class Msp:  
    def __init__(self):  
        pass  
  
    def login(self):  
        ret = dll.MSPLogin(None, None, login_params)  
        print(('MSPLogin =>'), ret)  
  
    def tts(self, text, filename, session_begin_params):  
        ret, audio_len, synth_status, getret, wav_datasize = c_int(), c_int(), c_int(), c_int(), c_int()   
        sessionID = dll.QTTSSessionBegin(session_begin_params, byref(ret))  
        ret = dll.QTTSTextPut(sessionID, text, len(text), None)  
  
        wavFile = open(filename, 'wb')  
        wavFile.write(wav_header)  
        while True:  
            data = dll.QTTSAudioGet(sessionID, byref(audio_len), byref(synth_status), byref(getret))  
            if audio_len.value>0:  
                print('datasize:            %d\r' % wav_datasize.value, end='\r')  
            if getret.value != MSP_SUCCESS:  
                break  
            if data:  
                wavFile.write(string_at(data, audio_len))  
                wav_datasize.value += audio_len.value  
            if synth_status.value == MSP_TTS_FLAG_DATA_END:  
                break  
            #time.sleep(0.01)  
        # fix wav header  
        WAVE_HEADER_SIZE = 44  
        if wav_datasize.value>0:  
            wav_size8 = c_int()  
            wav_size8.value = WAVE_HEADER_SIZE + wav_datasize.value - 8  
            wavFile.seek(4)  
            wavFile.write(wav_size8)  
            wavFile.seek(40)  
            wavFile.write(wav_datasize)  
        wavFile.close()  
        dll.QTTSSessionEnd(sessionID, "Normal")  
  
if __name__ == "__main__":  
    text = "欢迎来到方工的CSDN博客。CSDN创立于1999年，是中国专业的IT技术社区和开发者服务平台。拥有5000万注册用户以及60万注册企业及合作伙伴。"  
    print(text)  
    msp = Msp()  
    msp.login()  
    msp.tts(text.encode('U8'), filename, session_begin_params)  

tts3.py
#! python2
#-*- coding: utf-8 -*-  
import requests  
import time  
import hashlib  
import base64  
  
URL = "http://api.xfyun.cn/v1/service/v1/tts"  
APPID = "5ade9bcc"  
API_KEY = "f24968286f853d6138ff1fda9f1bce10"  
def getHeader(auf, aue, voiceName, speed, volume, pitch, engineType, textType):  
    curTime = str(int(time.time()))  
    param = "{\"auf\":\""+auf+"\""  
    if aue != "":  
        param +=",\"aue\":\"" + aue + "\""  
  
    if voiceName != "":  
        param +=",\"voice_name\":\"" + voiceName + "\""  
  
    if speed != "":  
        param +=",\"speed\":\"" + speed + "\""  
  
    if volume != "":  
        param +=",\"volume\":\"" + volume + "\""  
  
    if pitch != "":  
        param +=",\"pitch\":\"" + pitch + "\""  
  
    if engineType != "":  
        param +=",\"engine_type\":\"" + engineType + "\""  
  
    if textType != "":  
        param +=",\"text_type\":\"" + textType + "\""  
  
    param +="}"  
  
    paramBase64 = base64.b64encode(param)  
  
    m2 = hashlib.md5()  
    m2.update(API_KEY + curTime + paramBase64)  
    checkSum = m2.hexdigest()  
    header ={  
        'X-CurTime':curTime,  
        'X-Param':paramBase64,  
        'X-Appid':APPID,  
        'X-CheckSum':checkSum,  
        'Content-Type':'application/x-www-form-urlencoded; charset=utf-8',  
    }  
    return header  
  
def getBody(text):  
    data = {'text':text}  
    return data  
  
def writeFile(file, content):  
    with open(file, 'wb') as f:  
        f.write(content)  
    f.close()  
  
r = requests.post(URL,headers=getHeader("audio/L16;rate=16000", "raw", "xiaoyan", "50", "50", "50", "aisound", "text"),data=getBody("请输入你的身高和体重"))  
contentType = r.headers['Content-Type']  
if contentType == "audio/mpeg":  
    sid = r.headers['sid']  
    writeFile("audio/"+sid+".wav", r.content)  
    print "success, sid = " + sid  
else :  
    print r.text  

word.py
#!/usr/bin/python
#! python2
# -*- coding: UTF-8 -*-
import urllib2
import time
import urllib
import json
import hashlib
import base64


def main():
    api_key = '2a5c520ae43bfa103043f11740933ef3'
    x_appid = '5ade9bcc'
    #AUDIO_PATH = 'D:/python/545986818.872027.mp3'
    #AUDIO_PATH = 'D:/python/output.mp3'
    AUDIO_PATH = 'D:/python/123.wav'
    max_wanted_size = 1024 * 1024
    f = open(AUDIO_PATH, 'rb')
    file_content = f.read(max_wanted_size)
    base64_audio = base64.b64encode(file_content)
    body = urllib.urlencode({'audio': base64_audio})

    url = 'http://api.xfyun.cn/v1/service/v1/iat'
    param = {"engine_type": "sms16k", "aue": "raw"}

    x_param = base64.b64encode(json.dumps(param).replace(' ', ''))
    x_time = int(int(round(time.time() * 1000)) / 1000)
    x_checksum = hashlib.md5(api_key + str(x_time) + x_param).hexdigest()
    x_header = {'X-Appid': x_appid,
                'X-CurTime': x_time,
                'X-Param': x_param,
                'X-CheckSum': x_checksum,
                'Content-Type':'application/x-www-form-urlencoded; charset=utf-8',  
               }
    req = urllib2.Request(url, body, x_header)
    result = urllib2.urlopen(req)
    result = result.read()
    print result
    news = open('news.txt','wb')
    print(news.write(result))    
    return

if __name__ == '__main__':
    main() 

scrap1.py
from urllib.request import urlopen
from bs4 import BeautifulSoup
import csv
from datetime import datetime

'''
quote_page = 'http://www.bloomberg.com/quote/SPX:IND'
page = urlopen(quote_page)
soup = BeautifulSoup(page, 'html.parser')
name_box = soup.find('h1', attrs={'class': 'name'})
name = name_box.text.strip()
print (name)
# strip() 函數用於去除前後空格
price_box = soup.find('div', attrs={'class':'price'})
price = price_box.text
print (price)

with open('index.csv', 'a') as csv_file:
	writer = csv.writer(csv_file)
	writer.writerow([name, price, datetime.now()])
'''

quote_page = ['http://www.bloomberg.com/quote/SPX:IND', 'http://www.bloomberg.com/quote/CCMP:IND']
data = []
for pg in quote_page:
 	page = urlopen(pg)
 	soup = BeautifulSoup(page, 'html.parser')
name_box = soup.find('h1', attrs={'class': 'name'})
name = name_box.text.strip() # strip() is used to remove starting and trailing  
price_box = soup.find('div', attrs={'class':'price'})
price = price_box.text
data.append((name, price))

with open('index.csv', 'a') as csv_file:
	writer = csv.writer(csv_file)
	for name, price in data:
		writer.writerow([name, price, datetime.now()])

word5.py
#影音轉文字
#! python2
# -*- coding: UTF-8 -*-  
import requests  
import time  
import urllib  
import json  
import hashlib  
import base64  
  
URL = "http://api.xfyun.cn/v1/service/v1/iat"  
APPID = "5ade9bcc"  
API_KEY = "2a5c520ae43bfa103043f11740933ef3"  
  
def getHeader():  
    curTime = str(int(time.time()))  
    param = "{\"engine_type\": \"sms16k\", \"aue\": \"raw\"}"  
    paramBase64 = base64.b64encode(param)  
  
    m2 = hashlib.md5()  
    m2.update(API_KEY + curTime + paramBase64)  
    checkSum = m2.hexdigest()  
    header ={  
        'X-CurTime':curTime,  
        'X-Param':paramBase64,  
        'X-Appid':APPID,  
        'X-CheckSum':checkSum,  
        'Content-Type':'application/x-www-form-urlencoded; charset=utf-8',  
    }  
    return header  
  
def main():  
    #filename = 'd:/python/545986818.872027.mp3'
    filename = '123.wav'
    f = open(filename, 'rb')  
    file_content = f.read()  
    base64_audio = base64.b64encode(file_content)  
    body = urllib.urlencode({'audio': base64_audio})  

    r = requests.post(URL,headers=getHeader(),data=body)  
    result = json.loads(r.content)  
  
    if result["code"] == "0":  
        print "success, data = " + result["data"]  
    else:  
        print r.text  
  
    return  
  
if __name__ == '__main__':  
    main()  